src/ring_theory/witt_vector/frobenius_fraction_field.lean

```diff
@@ -0,0 +1,587 @@
+  solution p a₁ a₂ ≠ 0 :=
+begin
+  intro h,
+  have := solution_spec p a₁ a₂,
+  rw [h, zero_pow] at this,
+  { simpa [ha₁, ha₂] using _root_.div_eq_zero_iff.mp this.symm },
+  { linarith [hp.out.one_lt, le_of_lt hp.out.one_lt] }
+end
+
+lemma solution_spec' {a₁ : 𝕎 k} (ha₁ : a₁.coeff 0 ≠ 0) (a₂ : 𝕎 k) :
+  (solution p a₁ a₂)^p * a₁.coeff 0 = (solution p a₁ a₂) * a₂.coeff 0 :=
+begin
+  have := solution_spec p a₁ a₂,
+  cases nat.exists_eq_succ_of_ne_zero hp.out.ne_zero with q hq,
+  have hq' : q = p - 1 := by simp only [hq, tsub_zero, nat.succ_sub_succ_eq_sub],
+  conv_lhs {congr, congr, skip, rw hq},
+  rw [pow_succ', hq', this],
+  field_simp [ha₁, mul_comm],
+end
+
+end alg_closed_base
+open alg_closed_base
+
+section frobenius_rotation
+
+variables [is_alg_closed k] [char_p k p]
+
+/--
+Recursively defines the sequence of coefficients for `witt_vector.frobenius_rotation`.
+-/
+noncomputable def frobenius_rotation_coeff {a₁ a₂ : 𝕎 k}
+  (ha₁ : a₁.coeff 0 ≠ 0) (ha₂ : a₂.coeff 0 ≠ 0) : ℕ → k
+| 0       := solution p a₁ a₂
+| (n + 1) := succ_nth_val p n a₁ a₂ (λ i, frobenius_rotation_coeff i.val) ha₁ ha₂
+using_well_founded { dec_tac := `[apply fin.is_lt] }
+
+/--
+For nonzero `a₁` and `a₂`, `frobenius_rotation a₁ a₂` is a Witt vector that satisfies the
+equation `frobenius (frobenius_rotation a₁ a₂) * a₁ = (frobenius_rotation a₁ a₂) * a₂`.
+-/
+def frobenius_rotation {a₁ a₂ : 𝕎 k} (ha₁ : a₁.coeff 0 ≠ 0) (ha₂ : a₂.coeff 0 ≠ 0) : 𝕎 k :=
+witt_vector.mk p (frobenius_rotation_coeff p ha₁ ha₂)
+
+lemma frobenius_rotation_nonzero {a₁ a₂ : 𝕎 k} (ha₁ : a₁.coeff 0 ≠ 0) (ha₂ : a₂.coeff 0 ≠ 0) :
+  frobenius_rotation p ha₁ ha₂ ≠ 0 :=
+begin
+  intro h,
+  apply solution_nonzero p ha₁ ha₂,
+  simpa [← h, frobenius_rotation, frobenius_rotation_coeff] using witt_vector.zero_coeff p k 0
+end
+
+lemma frobenius_frobenius_rotation {a₁ a₂ : 𝕎 k} (ha₁ : a₁.coeff 0 ≠ 0) (ha₂ : a₂.coeff 0 ≠ 0) :
+  frobenius (frobenius_rotation p ha₁ ha₂) * a₁ = (frobenius_rotation p ha₁ ha₂) * a₂ :=
+begin
+  ext n,
+  induction n with n ih,
+  { simp only [witt_vector.mul_coeff_zero, witt_vector.coeff_frobenius_char_p,
+      frobenius_rotation, frobenius_rotation_coeff],
+    apply solution_spec' _ ha₁ },
+  { simp only [nth_remainder_spec, witt_vector.coeff_frobenius_char_p, frobenius_rotation_coeff,
+      frobenius_rotation, fin.val_eq_coe],
+    have := succ_nth_val_spec' p n a₁ a₂
+      (λ (i : fin (n + 1)), frobenius_rotation_coeff p ha₁ ha₂ i.val) ha₁ ha₂,
+    simp only [frobenius_rotation_coeff, fin.val_eq_coe, fin.val_zero] at this,
+    convert this using 4,
+    apply truncated_witt_vector.ext,
+    intro i,
+    simp only [fin.val_eq_coe, witt_vector.coeff_truncate_fun, witt_vector.coeff_frobenius_char_p],
+    refl }
+end
+
+local notation `K` := fraction_ring (𝕎 k)
+
+/-- This is basically the same as `𝕎 k` being a DVR. -/
+lemma exists_eq_pow_p_mul (a : 𝕎 k) (ha : a ≠ 0) :
+  ∃ (m : ℕ) (b : 𝕎 k), b.coeff 0 ≠ 0 ∧ a = p ^ m * b :=
```

Comment:
I wonder if this lemma should move and/or be given different assumptions -- I don't think it uses the algebraic closedness of `k`.

Reply:
I totally agree that we want this lemma in as much generality as possible (I think `k` a field of char `p` should be enough). Do we know that `b` can be chosen to be a unit?

Reply:
> Do we know that `b` can be chosen to be a unit?

I don't think mathlib knows that (under some hypotheses?) `b.coeff 0 ≠ 0` is equivalent to `b` being a unit -- if that's what you're asking?

Reply:
My question is if we can write any `a ≠ 0` as `a = p ^ m * u` where `u` is a unit (since this is what we do in a DVR). But if this is not already there I am not suggesting to add it now.

Reply:
> My question is if we can write any `a ≠ 0` as `a = p ^ m * u` where `u` is a unit (since this is what we do in a DVR). But if this is not already there I am not suggesting to add it now.

@riccardobrasca I added this variant.  It required slightly upgraded hypotheses (perfect field rather than perfect ring) but otherwise wasn't particularly hard (+60 lines).

In adding this variant I discovered what might be a useful division of this PR into smaller PRs and/or smaller files.  I had to re-use @robertylewis' lemma
```
lemma nth_mul_coeff (n : ℕ) :
  ∃ f : (truncated_witt_vector p (n+1) k → truncated_witt_vector p (n+1) k → k), ∀ (x y : 𝕎 k),
    (x * y).coeff (n+1) =
      x.coeff (n+1) * y.coeff 0 ^ (p^(n+1)) + y.coeff (n+1) * x.coeff 0 ^ (p^(n+1)) +
      f (truncate_fun (n+1) x) (truncate_fun (n+1) y) :=
```
which expreses the `n + 1`-th coefficient of a product of Witt vectors, modulo the earlier coefficients.  So maybe that lemma and the 300 (!) lines leading up to it are worth moving elsewhere -- they could even live in the file `witt_vector.truncated` if that doesn't seem too odd?

Reply:
Oh, nice, the generalization looks great!

I think putting them in `truncated` is weird. In fact, I don't think these use any of the structure on truncated Witt vectors beyond the fact that they're `n`-tuples. I might try to rephrase this in terms of `fin n → k`. But this could live in its own file.

Reply:
Very nice! This really says that Witt vectors are a DVR!

Reply:
@riccardobrasca Well, I tried to make them officially a `discrete_valuation_ring`, and got stuck because I didn't know how to prove `p` was `irreducible` :-)